libs/ml/include/ml/dataloaders/tensor_dataloader.hpp

```diff
@@ -99,12 +107,19 @@ TensorDataLoader<LabelType, InputType>::GetNext()
   {
     throw std::runtime_error("random mode not implemented for tensor dataloader");
   }
+
+  if (this->mode_ == DataLoaderMode::VALIDATE)
```

Comment:
a single if/else is not very expensive, but it is happening inside of the GetNext() method, which is very time critical. We can leave this optimisation for another PR, but I'll point it out here:
Instead of SetMode(new_mode) switching the enum, and then having this if/else based on the enum, we could instead have a third cursor which is a pointer to either the validation or training cursor - Set mode would update the enum and make this third cursor point to either the validation or training data as necessary. Then inside GetNext() we have shorter code and no if/else statement


One way we could make these even faster is if when SetMode(new_mode) is called,